ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -765,6 +805,68 @@ module Winbind = struct
       hostname
 end
 
+module ClosestKdc = struct
+  let invalid_kdc = "invalid_kdc"
+  let periodic_update_task_name = "Update closest kdc"
+  let mtime_this ~name ~f =
+    try
+      let start = Mtime_clock.counter() in
+      f();
+      (name, Mtime_clock.count start)
+    with _ ->
+      debug "Failed to time %s" name;
+      (name, Mtime.Span.max_span)
+
+  let update_db ~domain ~kdc =
+  (fun __context ->
+     let self = Helpers.get_localhost ~__context in
+     Db.Host.get_external_auth_configuration ~__context ~self|> fun value ->
+     (domain, kdc) :: List.remove_assoc domain value |> fun value ->
+     Db.Host.set_external_auth_configuration ~__context ~self ~value)
+  |> Server_helpers.exec_with_new_task "update domain closest kdc"
+
+  let from_db domain =
+    (fun __context ->
+     let self = Helpers.get_localhost ~__context in
+     Db.Host.get_external_auth_configuration ~__context ~self|> List.assoc_opt domain)
+  |> Server_helpers.exec_with_new_task "query domain closest kdc"
+
+  let lookup domain =
+    try
+      let krbtgt_sid = Wbinfo.sid_of_name (Printf.sprintf "%s@%s" krbtgt domain) |> maybe_raise in
+      let domain_netbios_name = Wbinfo.domain_name_of ~target_name_type:NetbiosName ~from_name: domain |> maybe_raise in
+
+      kdcs_of_domain domain
+      |> List.map (fun kdc -> debug "Got domain '%s' kdc '%s'" domain kdc; kdc)
+      |> List.map (fun kdc -> mtime_this ~name:kdc ~f:(fun() -> Ldap.query_user krbtgt_sid domain_netbios_name kdc |> maybe_raise))
+      |> List.filter (fun (_, span) -> Mtime.Span.equal Mtime.Span.max_span span|> not)
+      |> List.sort (fun (_, s1) (_, s2) -> Mtime.Span.compare s1 s2)
+      |> hd (Printf.sprintf "domain %s does not has valid kdc" domain)
+      |> fst
+    with _ ->
+      debug "Failed to lookup domain %s closest kdc" domain;
+      invalid_kdc
```

Comment:
maybe you would do this in javascript, but not in ocaml :)

please use `Ok/Error` to encode whether the computation succeeded or not, rather than trying to encode all the information with a string